### PR TITLE
Implements output shape checks for t_batch_mode_transform

### PR DESCRIPTION
**Pull request by saitcakmak**
## Motivation

In most cases, the acquisition functions are intended to return an output with `shape` matching the `batch_shape` of `X`, or the `model._input_batch_shape` in the case of acquisition functions using batched models. In certain cases (#573), when the acqf attributes are mis-specified, we get an output of wrong shape, which either leads errors down the line, or could possibly lead to buggy output. This implements an `assert_output_shape` flag in `t_batch_mode_transform` (defaults to `True`), which checks the output shape to catch the error in place.

Other changes:
`qMultiFidelityKnowledgeGradient` was missing from `__init__.py`, added it.
Set `assert_output_shape=False` for `ExpectedImprovement`, since `NoisyExpectedImprovement` uses it to get output with an additional dimension.
Updated some mock tests to match the correct input / output shapes.

(Write your motivation here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)
Yes.

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

Added unit tests to ensure everything works as intended. This also lead to discovering some shape mismatches in the mock tests in `test.acquisition`, which got fixed in the process.

cc @Balandat

### CODE REVIEW

**Comment by saitcakmak** [comment]
That was an interesting failure. CircleCI fails black formatting that is enforced by my local black installation. I updated it to match what CircleCI recommends.

File: botorch/utils/transforms.py [hunk 1]
```diff
@@ -166,7 +171,29 @@ def decorated(cls: Any, X: Tensor, **kwargs: Any) -> Any:
...
+                and not (
+                    output.shape == torch.Size() and X.shape[:-2] == torch.Size([1])
+                )  # output of size 1 is typically squeezed
+            ):
+                # the check above may fail when using a batched model.
+                # ensure that this is not the case
+                from botorch.acquisition import AcquisitionFunction
+
+                if (
+                    not isinstance(cls, AcquisitionFunction)
```
Comment (Balandat) [inline botorch/utils/transforms.py:187]: Do we ever use this decorator outside of acquisition functions?

File: botorch/utils/transforms.py [hunk 2]
```diff
@@ -166,7 +171,29 @@ def decorated(cls: Any, X: Tensor, **kwargs: Any) -> Any:
...
+                    or output.shape != cls.model._input_batch_shape
```
Comment (Balandat) [inline botorch/utils/transforms.py:188]: this may raise an `AttributeError` for models that don't have an `_input_batch_shape` property.

**Review by Balandat — COMMENTED** [review]
Thanks for putting this up and for adding solid tests. The fact that we can have batched models makes the shape checking somewhat complicated. Let me think a little bit about what the best way of doing this in a general setting is. 

Maybe @sdaulton has some thoughts on this as well.

File: botorch/utils/transforms.py (hunk 1 above)
Comment (saitcakmak) [reply botorch/utils/transforms.py:187]: Looks like it is currently only used for acquisition functions. I'll remove that check

**Comment by saitcakmak** [comment]
Thanks for reviewing this. I updated the handling of batched models in line with the comments. 

My experience with batched models is pretty much limited to fantasy models, so I don't know if there are cases this is missing. I can update it / add tests later based on the feedback.

**Comment by saitcakmak** [comment]
@Balandat I'll bring this up to date with #588. Are there other changes this needs?

File: botorch/utils/transforms.py [hunk 3]
```diff
@@ -166,7 +171,25 @@ def decorated(cls: Any, X: Tensor, **kwargs: Any) -> Any:
                     f" got X with shape {X.shape}."
                 )
             X = X if X.dim() > 2 else X.unsqueeze(0)
-            return method(cls, X, **kwargs)
+            output = method(cls, X, *args, **kwargs)
+            if (
```
Comment (Balandat) [inline botorch/utils/transforms.py:175]: This whole if clause is kind of complex, mind putting this into a helper function that returns a bool? That will clean things up. 

Also, can you migrate this from using the `_input_batch_shape` to the universal `batch_shape` interface introduced in #588?

**Review by Balandat — CHANGES_REQUESTED** [review]

**Comment by saitcakmak** [comment]
I made the changes and added & fixed some test cases.

File: botorch/utils/transforms.py [hunk 4]
```diff
@@ -124,19 +124,52 @@ def normalize_indices(indices: Optional[List[int]], d: int) -> Optional[List[int
     return normalized_indices
 
 
+def _verify_output_shape(cls: Any, X: Tensor, output: Tensor) -> bool:
```
Comment (Balandat) [inline botorch/utils/transforms.py:127]: If the first argument is an acquisition function object, let's maybe use `acqf: AcquisitionFunction` for this?

File: botorch/utils/transforms.py [hunk 5]
```diff
@@ -166,7 +199,17 @@ def decorated(cls: Any, X: Tensor, **kwargs: Any) -> Any:
...
                 )
             X = X if X.dim() > 2 else X.unsqueeze(0)
-            return method(cls, X, **kwargs)
+            output = method(cls, X, *args, **kwargs)
+            if assert_output_shape and not _verify_output_shape(
+                cls=cls, X=X, output=output
+            ):
+                raise AssertionError(
+                    "Expected the output shape to match either the batch shape of "
+                    "X, or the `model._input_batch_shape` in the case of "
```
Comment (Balandat) [inline botorch/utils/transforms.py:208]:
```suggestion
                    "X, or the `model.batch_shape` in the case of "
```

File: botorch/utils/transforms.py [hunk 6]
```diff
@@ -124,19 +124,52 @@ def normalize_indices(indices: Optional[List[int]], d: int) -> Optional[List[int
...
+    """
+    try:
+        return (
+            output.shape == X.shape[:-2]
+            or (output.shape == torch.Size() and X.shape[:-2] == torch.Size([1]))
+            or output.shape == cls.model.batch_shape
+        )
+    except (AttributeError, NotImplementedError):
+        # cls does not have model or cls.model does not define `batch_shape`
+        return False
```
Comment (Balandat) [inline botorch/utils/transforms.py:152]: So I don't know if returning `False` here is the right thing to do. Basically if we get to this point, we have made some assumptions about the acquisition function (that it operates on a model that implements a `batch_shape` property) that are not true. I think maybe being somewhat lenient here and returning `True` but raising a `RuntimeWarning` pointing this out instead may be the better approach (as in less annoying to work around if doing custom stuff)?

File: test/acquisition/test_knowledge_gradient.py [hunk 7]
```diff
@@ -349,7 +349,9 @@ def test_evaluate_q_multi_fidelity_knowledge_gradient(self):
                     patch_f.assert_called_once()
                     cargs, ckwargs = patch_f.call_args
                     self.assertEqual(ckwargs["X"].shape, torch.Size([b, 1, 1]))
-            val_exp = mock_util(X, mean.squeeze(-1) - current_value).mean(dim=0)
+            val_exp = mock_util(X, mean.squeeze(-1).squeeze(-1) - current_value).mean(
```
Comment (Balandat) [inline test/acquisition/test_knowledge_gradient.py:352]:
```suggestion
            val_exp = mock_util(X, mean.view(mean.shape[:-2]) - current_value).mean(
```

File: test/acquisition/test_monte_carlo.py [hunk 8]
```diff
@@ -761,7 +761,7 @@ def test_q_upper_confidence_bound_batch(self):
             acqf.set_X_pending(None)
             self.assertIsNone(acqf.X_pending)
             acqf.set_X_pending(X)
-            self.assertEqual(acqf.X_pending, X)
+            self.assertTrue((acqf.X_pending == X).all())
```
Comment (Balandat) [inline test/acquisition/test_monte_carlo.py:764]:
```suggestion
            self.assertTrue(torch.equal(acqf.X_pending, X))
```

**Review by Balandat — CHANGES_REQUESTED** [review]
Looks mostly good to me, thanks for polishing this.

My main comment is on how to handle the case when the acquisition function doesn't have a model or if the model doesn't implement `batch_shape`.

The rest are mostly nits.

File: test/utils/test_transforms.py [hunk 9]
```diff
@@ -17,7 +17,10 @@
     standardize,
     t_batch_mode_transform,
     unnormalize,
+    _verify_output_shape,
 )
+from botorch.models.model import Model
+from botorch.utils.testing import MockModel, MockPosterior
```
Comment (Balandat) [inline test/utils/test_transforms.py:23]: oh and can you rearrange these inputs alphabetically? linter will complain about this.

Comment (saitcakmak) [reply test/utils/test_transforms.py:23]: What convention do we follow for ordering inputs? Using the notation `from <package> import <module>`, this is currently sorted for the <package> and over <module> for each package. Does linter expect it to be sorted over <module> instead? Trying to get a better understanding for future reference

File: botorch/utils/transforms.py (hunk 6 above)
Comment (saitcakmak) [reply botorch/utils/transforms.py:152]: I am thinking in what scenario you'd get to here. Either you'd use one of the canned acquisition functions (that uses `t_batch_mode_transform`) with a custom model that doesn't define `batch_shape`, or you wrote a custom class, used `t_batch_model_transform` in it without passing `assert_output_shape = False`. In the second scenario, it is easy to fix by passing `assert_output_shape = False`, but the first one would require editing the source or redefining the acquisition function, which is annoying. Yeah, ignoring a `RuntimeWarning` sounds less annoying.

File: botorch/utils/transforms.py [hunk 10]
```diff
@@ -154,7 +187,7 @@ def t_batch_mode_transform(
 
     def decorator(method: Callable[[Any, Tensor], Any]) -> Callable[[Any, Tensor], Any]:
         @wraps(method)
-        def decorated(cls: Any, X: Tensor, **kwargs: Any) -> Any:
+        def decorated(cls: Any, X: Tensor, *args: Any, **kwargs: Any) -> Any:
```
Comment (saitcakmak) [inline botorch/utils/transforms.py:190]: @Balandat, Should this also change to `acqf: AcquisitionFunction` since that is the only use case? I think it makes sense to keep the signature consistent with `_verify_output_shape`.

File: test/utils/test_transforms.py (hunk 9 above)
Comment (Balandat) [reply test/utils/test_transforms.py:23]: So we have been using iSort so far - I'm about to switch over to usort in #603, but there is one edge case that has to do with FB-infra that needs to be fixed upstream in usort before I can do that. 

Basically to fix this you just need to merge the `botorch.utils.testing` imports with the existing one above and move the models import up to be in alphabetic order.

File: botorch/utils/transforms.py (hunk 4 above)
Comment (saitcakmak) [reply botorch/utils/transforms.py:127]: This actually leads to a circular import
